### PR TITLE
Fix Iceberg read when positional delete files are unaligned

### PR DESCRIPTION
**Pull request by yingsu00**
When the base data file and positional delete files contains multiple
unaligned RowGroups, some of the bits at the end of
IcebergSplitReader::deleteBitmap_ could be mistakenly skipped, causing
wrong result. This commit fixes it by introducing an offset into this
deleteBitmap_ and shift the unused bits to the beginning for each batch.

Fixes https://github.com/facebookincubator/velox/issues/9856

### CODE REVIEW

File: velox/connectors/hive/iceberg/IcebergSplitReader.cpp [hunk 1]
```diff
@@ -97,31 +99,74 @@ uint64_t IcebergSplitReader::next(uint64_t size, VectorPtr& output) {
...
+      deleteBitmap_ ? deleteBitmap_->as<uint64_t>() : nullptr;
+
   auto rowsScanned = baseRowReader_->next(size, output, &mutation);
   baseReadOffset_ += rowsScanned;
+  deleteBitmapBitOffset_ = rowsScanned;
 
   return rowsScanned;
 }
 
+void IcebergSplitReader::shiftDeleteBitmap() {
```
[the remark this answers is not in the document]
Comment (yingsu00) [reply velox/connectors/hive/iceberg/IcebergSplitReader.cpp]: Thanks Jimmy! I updated to bits::copyBits(). 

btw. Just out of curiosity, I benchmarked bits::copyBits() and this shiftDeleteBitmap () function, and found bits::copyBits() was over 10x slower than this for many cases. Maybe we can improve the bits operations in the future. 
Screenshot 2024-06-21 at 10 16 09

File: velox/connectors/hive/iceberg/IcebergSplitReader.cpp [hunk 2]
```diff
@@ -97,31 +99,74 @@ uint64_t IcebergSplitReader::next(uint64_t size, VectorPtr& output) {
...
     }
-
-    deleteBitmap_->setSize(numBytes);
-    mutation.deletedRows = deleteBitmap_->as<uint64_t>();
   }
 
+  mutation.deletedRows =
+      deleteBitmap_ ? deleteBitmap_->as<uint64_t>() : nullptr;
+
   auto rowsScanned = baseRowReader_->next(size, output, &mutation);
```
[the remark this answers is not in the document]
Comment (yingsu00) [reply velox/connectors/hive/iceberg/IcebergSplitReader.cpp:147]: @Yuhta Thanks Jimmy! I checked these functions and yes we can get the actual rows scanned before calling base RowReader's next(). So I'll send a separate PR to simplify the logic using these functions. 
For now I updated the PR with your another comment addressed. Would appreciate your review again.

**Comment by yingsu00** [comment]
@Yuhta Would you please review again? Thank you!

File: velox/connectors/hive/iceberg/IcebergSplitReader.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Yuhta) [reply velox/connectors/hive/iceberg/IcebergSplitReader.cpp]: Yes there is definitely room to improve `copyBits`.  The overall timing of that function in the query is not large though.

**Review by Yuhta — APPROVED** [review]

**Comment by yingsu00** [comment]
@Yuhta Thanks for reviewing and approving this PR. Do we know when it will be merged?

File: velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp [hunk 3]
```diff
@@ -34,107 +34,186 @@ namespace facebook::velox::connector::hive::iceberg {
...
+  void assertSingleBaseFileSingleDeleteFile(
+      const std::vector<int64_t>& deletePositionsVec) {
+    std::map<std::string, std::vector<int64_t>> rowGroupSizesForFiles = {
+        {"data_file_1", {10000, 10000}}};
+    std::unordered_map<
+        std::string,
+        std::multimap<std::string, std::vector<int64_t>>>
+        deleteFilesForBaseDatafiles = {
+            {"delete_file_1", {{"data_file_1", deletePositionsVec}}}};
+    false;
```
Comment (Yuhta) [inline velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp:58]: Remove this

File: velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp [hunk 4]
```diff
@@ -34,107 +34,186 @@ namespace facebook::velox::connector::hive::iceberg {
...
+  ///     }
+  /// represents one delete file called delete_file_1, which contains delete
+  /// positions for data_file_1 and data_file_2. THere are 3 RowGroups in this
+  /// delete file, the first two contain positions for data_file_1, and the last
+  /// contain positions for data_file_2
+  void assertPositionalDeletes(
+      const std::map<std::string, std::vector<int64_t>>& rowGroupSizesForFiles,
+      const std::unordered_map<
+          std::string,
+          std::multimap<std::string, std::vector<int64_t>>>
```
Comment (Yuhta) [inline velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp:170]: Pass by reference

File: velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp [hunk 5]
```diff
@@ -146,6 +225,116 @@ class HiveIcebergTest : public HiveConnectorTestBase {
...
+  /// <"dataFile2", {pos_RG1, pos_RG2,..}>
+  std::unordered_map<
+      std::string,
+      std::pair<int64_t, std::shared_ptr<TempFilePath>>>
+  writePositionDeleteFiles(
+      const std::unordered_map<
+          std::string, // delete file name
+          std::multimap<
+              std::string,
+              std::vector<int64_t>>>
```
Comment (Yuhta) [inline velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp:271]: Ditto

File: velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp [hunk 6]
```diff
@@ -170,106 +359,114 @@ class HiveIcebergTest : public HiveConnectorTestBase {
...
+          std::multimap<std::string, std::vector<int64_t>>>
+          deleteFilesForBaseDatafiles) {
+    int64_t totalNumRowsInAllBaseFiles = 0;
+    std::map<std::string, int64_t> baseFileSizes;
+    for (auto rowGroupSizesInFile : rowGroupSizesForFiles) {
+      // Sum up the row counts in all RowGroups in each base file
+      baseFileSizes[rowGroupSizesInFile.first] += std::accumulate(
+          rowGroupSizesInFile.second.begin(),
+          rowGroupSizesInFile.second.end(),
+          0);
```
Comment (Yuhta) [inline velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp:375]: `0LL`

**Review by Yuhta — COMMENTED** [review]
Some lint errors

**Comment by yingsu00** [comment]
@Yuhta Thanks for importing the PR and let me know the failures! I have updated the PR with your comments addressed.

**Pull request closed without merging**